### PR TITLE
std: Fix file locking logic for BSD targets

### PR DESCRIPTION
**Pull request by LemonBoy**
Spotted while playing with a FreeBSD machine.

@semarie This affects the OpenBSD codepath too, is this working for you too?

### CODE REVIEW

**Comment by LemonBoy** [comment]
So.. using `O_SHLOCK` (or its exclusive variant) forces the file to be opened in no blocking mode!
This is definitely an unwanted side effect, we only need the non blocking bit during the open.

**Comment by daurnimator** [comment]
>  This is definitely an unwanted side effect, we only need the non blocking bit during the open.

Then after the open, remove it via an `fcntl` call?

**Comment by semarie** [comment]
> So.. using O_SHLOCK (or its exclusive variant) forces the file to be opened in no blocking mode!

I don't see such behaviour in [open(2)](https://www.freebsd.org/cgi/man.cgi?query=open&apropos=0&sektion=2&manpath=FreeBSD+12.2-RELEASE+and+Ports&arch=default&format=html) man page.

Could you check with [ktrace(1)](https://www.freebsd.org/cgi/man.cgi?query=ktrace&apropos=0&sektion=0&manpath=FreeBSD+12.2-RELEASE+and+Ports&arch=default&format=html) that the `open()` syscall is used with the expected flags ? (`ktrace` command-line utility is a bit like `strace` in linux world. run `ktrace program` to record the sycalls used by `program`, and next run `kdump` to show the recorded trace)

**Comment by LemonBoy** [comment]
> > So.. using O_SHLOCK (or its exclusive variant) forces the file to be opened in no blocking mode!
> 
> I don't see such behaviour in [open(2)](https://www.freebsd.org/cgi/man.cgi?query=open&apropos=0&sektion=2&manpath=FreeBSD+12.2-RELEASE+and+Ports&arch=default&format=html) man page.
> 
> Could you check with [ktrace(1)](https://www.freebsd.org/cgi/man.cgi?query=ktrace&apropos=0&sektion=0&manpath=FreeBSD+12.2-RELEASE+and+Ports&arch=default&format=html) that the `open()` syscall is used with the expected flags ? (`ktrace` command-line utility is a bit like `strace` in linux world. run `ktrace program` to record the sycalls used by `program`, and next run `kdump` to show the recorded trace)

It's the zig code that enables the `o_nonblock` bit when trying to obtain a lock, the behaviour we want here is described in the man page:

> [EWOULDBLOCK]	O_NONBLOCK and one of O_SHLOCK or O_EXLOCK is speci-
			fied and the file is locked.

**Comment by semarie** [comment]
I am unsure that you will be able to get a behaviour respecting both `lock_nonblocking=true` and `intented_io_mode=.blocking` with only one syscall.

as @daurnimator noted, you could remove it via `fcntl(2)`.

alternatively, you could open the file without locking, and if locking is desired use `flock(2)` on the descriptor. You should be able to block or not, depending the flags used for `flock(2)` (`LOCK_SH` or `LOCK_SH | LOCK_NB`)

File: lib/std/fs.zig [hunk 1]
```diff
@@ -771,6 +771,22 @@ pub const Dir = struct {
             });
         }
 
+        if (has_flock_open_flags and flags.lock_nonblocking) {
+            var fl_flags = os.fcntl(fd, os.F_GETFL, 0) catch |err| switch (err) {
```
Comment (daurnimator) [inline lib/std/fs.zig:775]: There should be no need to get fcntl flags for a freshly opened file descriptor.

**Comment by LemonBoy** [comment]
Let's take a step back, I asked if you had noticed any weirdness on OpenBSD as the file locking tests deadlock on FreeBSD without this patch.

> as @daurnimator noted, you could remove it via fcntl(2).

Yes, I think I know pretty well how to do that :) The problem was the missing `fcntl` call.

> alternatively, you could open the file without locking, and if locking is desired use flock(2) on the descriptor. You should be able to block or not, depending the flags used for flock(2) (LOCK_SH or LOCK_SH | LOCK_NB)

We try to use the `open` flags when available as then obtaining a lock is an atomic operation, the two step process with `open` and `flock` is prone to race conditions but is all we have on Linux.

File: lib/std/fs.zig (hunk 1 above)
Comment (LemonBoy) [reply lib/std/fs.zig:775]: I wish it was this easy, on a Linux system you can see there's at least something like `O_WRONLY|O_LARGEFILE`.

**Comment by LemonBoy** [comment]
@leroycep I think this is the root cause of the deadlock you were seeing on macos.
I didn't enable it back because I couldn't understand the need for `O_SYNC` in the `open` flags, do you mind explaining why it's needed here?

**Comment by semarie** [comment]
> Let's take a step back, I asked if you had noticed any weirdness on OpenBSD as the file locking tests deadlock on FreeBSD without this patch.

I think I will need some guidance to test only that. Currently, I am unsure about what I am running as tests.

```
$ stage1/zig build test -Dskip-non-native -Dtest-filter='lock'
Test [1/25] test "self-hosted"... tests [1/26] referencing decls which appear later in the file [1/1] update [1/3] writeAll 25 tests passed.
All 7 tests passed.
All 7 tests passed.
All 7 tests passed.
All 7 tests passed.
All 7 tests passed.
All 7 tests passed.
All 7 tests passed.
All 7 tests passed.
All 1 tests passed.
Semantic Analysis [1025/1187] No tests to run.
All 1 tests passed.
Semantic Analysis [940/1072] No tests to run.
All 1 tests passed.
Semantic Analysis [1018/1178] No tests to run.
All 1 tests passed.
Semantic Analysis [940/1072] No tests to run.
Semantic Analysis [1420/1585] No tests to run.
Semantic Analysis [1335/1470] No tests to run.
Semantic Analysis [1413/1576] No tests to run.
Semantic Analysis [1335/1470] No tests to run.
Test 1/3 compile-error suspend inside suspend block...OK
Test 2/3 compile-error implicit semicolon - block statement...OK
Test 3/3 compile-error implicit semicolon - block expr...OK
docgen example code 1/267...OK
docgen example code 2/267...OK
docgen example code 3/267...OK
docgen example code 4/267...OK
[...]
```

**Comment by LemonBoy** [comment]
The easier way is to run the fs tests individually `zig test --test-filter "lock lib/std/fs.zig`, specifying `--verbose-link` lets you see the binary name to run it under `truss` or `ktrace` or whatever.

**Comment by semarie** [comment]
so without the patch, "open file with exclusive nonblocking lock twice" test is hanging.

```
  2625 test     CALL  openat(6,0x7f7ffffd3740,0x10621<O_WRONLY|O_EXLOCK|O_CREAT|O_TRUNC|O_CLOEXEC>,0666<S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH>)
  2625 test     NAMI  "file_nonblocking_lock_test.txt"
  2625 test     RET   openat 4
  2625 test     CALL  openat(6,0x7f7ffffd3740,0x10621<O_WRONLY|O_EXLOCK|O_CREAT|O_TRUNC|O_CLOEXEC>,0666<S_IRUSR|S_IWUSR|S_IRGRP|S_IWGRP|S_IROTH|S_IWOTH>)
  2625 test     NAMI  "file_nonblocking_lock_test.txt"
... waiting
```

from flags passed to `openat(2)` (`O_WRONLY|O_EXLOCK|O_CREAT|O_TRUNC|O_CLOEXEC`) the blocking behaviour is expected (as there is no `O_NONBLOCKING`)

With the patch applied, tests are passing without hang.

**Comment by andrewrk** [comment]
You can also use `-femit-bin=foo` to put the binary wherever you want

**Comment by leroycep** [comment]
> @leroycep I think this is the root cause of the deadlock you were seeing on macos.
> I didn't enable it back because I couldn't understand the need for `O_SYNC` in the `open` flags, do you mind explaining why it's needed here?
I didn't remember why that was in there, so I tried looking it up. I couldn't find anything. I'm honestly not sure why it's there, so I see no reason to keep it.

**Review by andrewrk — APPROVED** [review]
This looks good to me - is it ready to merge?

**Comment by LemonBoy** [comment]
Tests are green, you can smash the green button.

**Pull request merged**